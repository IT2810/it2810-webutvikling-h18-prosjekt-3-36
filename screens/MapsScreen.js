import React from 'react';
import { connect } from 'react-redux';
import { View, StyleSheet } from 'react-native';
import { MapView, Permissions, Location } from 'expo';

class MapsScreen extends React.Component {
  static navigationOptions = {
    title: 'Map View',
  };

  state = {
    // Før du har fetched første resultat fra gpsen så skal du bare være zoomer ut og se hele verdenskartet.
    mapRegion: { latitude: 0, longitude: 0, latitudeDelta: 100, longitudeDelta: 100 },
    currentLocationResult: null,
  };

  componentDidMount() {
    // Fetch gps koordinater med en gang componenten har loadet.
    this._getLocationAsyncAndSetRegion();
  }

  onRegionChangeComplete = (mapRegion) => {
    this.setState({ mapRegion });
  }

  _getLocationAsyncAndSetRegion = async () => {
    // Kjøres for å zoome inn på riktig sted i verden. Første gps lokasjonen som du får.
    let { status } = await Permissions.askAsync(Permissions.LOCATION);
    if (status !== 'granted') {
      this.setState({
        currentLocationResult: 'Permission to access location was denied'
      });
    }
    // Det er en bug på android telefoner som gjør at denne ikke funker på mobiler som har "device only" på gps mode valgt i settings.
    let location = await Location.getCurrentPositionAsync({ enableHighAccuracy: false, maximumAge: 10000 });

    const lat = location["coords"]["latitude"];
    const long = location["coords"]["longitude"];
    const latDelta = 0.0922;
    const longDelta = 0.0421;
    this.setState({ mapRegion: { latitude: lat, longitude: long, latitudeDelta: latDelta, longitudeDelta: longDelta } })

    // this.setState({ currentLocationResult: { coords: { "latitude": location["coords"]["latitude"], "longitude": location["coords"]["longitude"] } } });

  };

  render() {
    const categoryToColor = {
      'work': 'red',
      'school': 'blue',
      'fun': 'orange',
      'other': 'green'
    };

    return (
      <View style={styles.container}>
        <MapView
          style={StyleSheet.absoluteFill}
          region={this.state.mapRegion}
          onRegionChangeComplete={this.onRegionChangeComplete}
          showsUserLocation={true}
          >

          {this.props.todos
          // Filtrer ut de todosene som ikke har koordinater. 
          .filter(todo => todo["coords"] !== undefined)
          .map(todo => 
            <MapView.Marker 
              key={todo["id"]}
              coordinate={todo["coords"]}
              title={todo["category"]}
              description={todo["text"]}
              pinColor={categoryToColor[todo["category"]]}
            />
          )} 

        </MapView>
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
  }
});

export default connect(state => ({ todos: state.todos }))(MapsScreen);