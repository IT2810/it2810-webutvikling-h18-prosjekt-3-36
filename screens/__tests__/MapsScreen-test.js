import React from "react";
import ShallowRenderer from 'react-test-renderer/shallow';
import MapsScreen from "../MapsScreen.js";


it("MapsScreen Matches snapshot.", () => {
    const renderer = new ShallowRenderer();
    renderer.render(<MapsScreen />)
    const result = renderer.getRenderOutput();
    expect(result).toMatchSnapshot();
});
