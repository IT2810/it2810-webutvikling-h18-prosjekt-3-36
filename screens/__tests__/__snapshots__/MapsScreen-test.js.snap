// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MapsScreen Matches snapshot. 1`] = `
<Component
  style={
    Object {
      "flex": 1,
    }
  }
>
  <MapView
    onRegionChangeComplete={[Function]}
    region={
      Object {
        "latitude": 0,
        "latitudeDelta": 100,
        "longitude": 0,
        "longitudeDelta": 100,
      }
    }
    showsUserLocation={true}
    style={
      Object {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  >
    <MapMarker
      coordinate={
        Object {
          "latitude": 63.4133335,
          "longitude": 10.4120182,
        }
      }
      description="kjøp melk"
      pinColor="red"
      stopPropagation={false}
      title="work"
    />
    <MapMarker
      coordinate={
        Object {
          "latitude": 63.4223335,
          "longitude": 10.4020182,
        }
      }
      description="gjør webdev og balbajsklfjkalsdfjkla sdjklfja skldfj klasjdfkl jaklsdfjkla jsdklfj askldjfkl jasdklfj klasjdfk jaskldfjklasj klfjaskld fjklasdfjkl "
      pinColor="orange"
      stopPropagation={false}
      title="fun"
    />
    <MapMarker
      coordinate={
        Object {
          "latitude": 63.4533335,
          "longitude": 10.4420182,
        }
      }
      description="kjøp melk"
      pinColor="blue"
      stopPropagation={false}
      title="school"
    />
    <MapMarker
      coordinate={
        Object {
          "latitude": 63.4833335,
          "longitude": 10.4920182,
        }
      }
      description="gjør webdev og balbajsklfjkalsdfjkla sdjklfja skldfj klasjdfkl jaklsdfjkla jsdklfj askldjfkl jasdklfj klasjdfk jaskldfjklasj klfjaskld fjklasdfjkl "
      pinColor="green"
      stopPropagation={false}
      title="other"
    />
  </MapView>
</Component>
`;
